File: source/lecture_1/problem_set_1/cash/cash.c
#include <stdio.h>

int get_cents(void);
int calculate_quarters(int cents);
int calculate_dimes(int cents);
int calculate_nickels(int cents);
int calculate_pennies(int cents);

int main(void)
{
    // Ask how many cents the customer is owed
    int cents = get_cents();

    if (cents == 0)
    {
        printf("0\n");
        return 0;
    }

    // Calculate the number of quarters to give the customer
    int quarters = calculate_quarters(cents);
    cents = cents - quarters * 25;
    // printf("cents: %i\n", cents);

    // Calculate the number of dimes to give the customer
    int dimes = calculate_dimes(cents);
    cents = cents - dimes * 10;
    // printf("cents: %i\n", cents);

    // Calculate the number of nickels to give the customer
    int nickels = calculate_nickels(cents);
    cents = cents - nickels * 5;
    // printf("cents: %i\n", cents);

    // Calculate the number of pennies to give the customer
    int pennies = calculate_pennies(cents);
    cents = cents - pennies * 1;
    // printf("cents: %i\n", cents);

    // Sum coins
    int coins = quarters + dimes + nickels + pennies;

    // Print total number of coins to give the customer
    printf("%i\n", coins);
}

int get_cents(void)
{
    int input;
    do
    {
        printf("Change owed: ");
        if (scanf("%i", &input) != 1)
        {
            int c;
            /* read and ignore the rest of the line */
            while ((c = getchar()) != EOF && c != '\n')
                continue;
            if (c == EOF)
            {
                /* premature end of file */
                return 1;
            }
            input = -1;
        }

    } while (input < 0);
}

int calculate_quarters(int cents)
{
    return cents / 25;
}

int calculate_dimes(int cents)
{
    return cents / 10;
}

int calculate_nickels(int cents)
{
    return cents / 5;
}

int calculate_pennies(int cents)
{
    return cents / 1;
}

File: source/lecture_6/problem_set/sentimental-cash/cash.py
import sys

QUARTER_VALUE = 25
DIME_VALUE = 10
NICKEL_VALUE = 5
PENNY_VALUE = 1


def get_cents():
    while True:
        try:
            # Get input from user.
            # Convert input to float, round to 2 decimal places and multiply by 100.
            # Convert to int for easier calculation.
            input_cents = int(round(float(input("Change owed: ")), 2) * 100)
            if input_cents >= 0:
                return input_cents
            else:
                print("Invalid input")
                continue    
        except ValueError:
            # Handle non-integer input
            print("Invalid input")
            continue
        except KeyboardInterrupt:
            # Handle Ctrl+C
            print()
            print("Exiting...")
            sys.exit(1)
        except EOFError:
            # Handle Ctrl+D
            print()
            print("Exiting...")
            sys.exit(1)


def calculate_quarters(cents: int):
    return int(cents / QUARTER_VALUE)

def calculate_dimes(cents: int):
    return int(cents / DIME_VALUE)

def calculate_nickels(cents: int):
    return int(cents / NICKEL_VALUE)

def calculate_pennies(cents: int):
    return int(cents / PENNY_VALUE)

def main():
    # Ask how many cents the customer is owed
    cents = get_cents()

    if cents == 0:
        print("0")
        return

    # Calculate the number of quarters to give the customer
    quarters = calculate_quarters(cents)
    cents = cents - quarters * QUARTER_VALUE

    # Calculate the number of dimes to give the customer
    dimes = calculate_dimes(cents)
    cents = cents - dimes * DIME_VALUE

    # Calculate the number of nickels to give the customer
    nickels = calculate_nickels(cents)
    cents = cents - nickels * NICKEL_VALUE

    # Calculate the number of pennies to give the customer
    pennies = calculate_pennies(cents)
    cents = cents - pennies * PENNY_VALUE

    # Sum coins
    coins = quarters + dimes + nickels + pennies

    # Print total number of coins to give the customer
    print(coins)

if __name__ == "__main__":
    main()
